### PR TITLE
Replace C style hash::lookup implementation with a more idiomatic Rust version.

### PR DESCRIPTION
**Pull request by shaleh**
Add a `HashLookupResult` which defines `Missing` and `Found`.

### CODE REVIEW

File: rust_src/src/hashtable.rs [hunk 1]
```diff
@@ -207,12 +221,9 @@ pub fn copy_hash_table(mut table: LispHashTableRef) -> LispHashTableRef {
...
-    let idx = hash_table.lookup(key, ptr::null_mut());
-
-    if idx >= 0 {
-        hash_table.get_hash_value(idx)
-    } else {
-        dflt
+    match hash_table.lookup(key) {
+        Found(idx) => hash_table.get_hash_value(idx),
+        Missing(_) => dflt,
     }
```
Comment (db48x) [inline rust_src/src/hashtable.rs]: I love this

Comment (shaleh) [reply rust_src/src/hashtable.rs]: Yeah, I pointed @brotzeit at it and then realized how poor the UX was.

Comment (shaleh) [reply rust_src/src/hashtable.rs]: I suppose we could get fancy and add `and_then` or `or_else` to it.

**Comment by db48x** [comment]
Heh

**Pull request merged**